design/src/rfcs/rfc0006_crates_io_publishing.md

```diff
@@ -0,0 +1,279 @@
+RFC: Publishing to Crates.io
+============================
+
+> Status: RFC
+
+The AWS SDK for Rust and its supporting Smithy crates need to be published to [crates.io](https://crates.io/)
+so that customers can include them in their projects and also publish crates of their own that depend on them.
+
+This doc proposes a short-term and long-term solution for publishing to crates.io. The short-term approach is intended
+to be executed manually by a developer using scripts and an SOP no more than once per week, and should require less
+than a dev week to implement. The long-term approach is intended to be fully automated and run as frequently as
+once per day.
+
+Terminology
+-----------
+
+- **AWS SDK Crate**: A crate that provides a client for calling a given AWS service, such as `aws-sdk-s3` for calling S3.
+- **AWS Runtime Crate**: Any runtime crate that the AWS SDK generated code relies on, such as `aws-types`.
+- **Smithy Runtime Crate**: Any runtime crate that the smithy-rs generated code relies on, such as `smithy-types`.
+
+Requirements
+------------
+
+### Versioning
+
+Cargo uses [semver](https://github.com/dtolnay/semver#requirements) for versioning,
+with a `major.minor.patch-pre` format:
+- `major`: Incompatible API changes
+- `minor`: Added functionality in backwards compatible manner
+- `patch`: Backwards compatible bug fixes
+- `pre`: Pre-release version tag (omitted for normal releases)
+
+The Rust SDK will make one small adjustment to the above requirements: the `patch` version will be incremented
+for backwards compatible AWS model changes, and the `minor` version will be incremented for changes in Smithy runtime
```

Comment:
Define "change" Is the change backwards compatible? Minor versions cannot be backwards incompatible.

Reply:
Tokio takes the approach that only significant bugs are fixed in patch releases, otherwise they are included in the next minor release. We release minor versions regularly, so it tends to be OK. We take the point of view that patch releases should be as stable as possible, so only introduce a patch release to fix important things.

Reply:
The only changes we would allow into a patch release are generated code from an updated service model. For example, if an AWS service adds a new API call, we would have some automation that pulls in the model update for that new API call and then generates a new SDK with it. This should only introduce new bugs if there is a bug in the code generator that gets exercised by the updated model (which is definitely possible).

The idea was that this kind of change could be a patch release since the model update is required to be backwards compatible, and thus, the code generated from it will also be backwards compatible. In reality, model updates aren't backwards compatible 100% of the time though, so an argument for putting them in minor releases rather than patches is valid.

Keeping the major/minor consistent across the SDK crates has the nice benefit that it's trivial to know that the correct versions are being used in a project. If we move model updates into minor, then I think we lose that unless we publish minor versions with no changes, or maintain some kind of version reference table for each SDK release.